blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md

```diff
@@ -0,0 +1,246 @@
+    Title: Lexical and Dynamic Scope
+    Date: 2019-08-14T13:59:21
+    Tags: scope, definitions, history, by Ming-Ho Yee
+
+This all started with a simple question about the R programming language: _is R
+lexically or dynamically scoped?_
+
+To answer that question, we need to understand what _scope_ is, along with
+_lexical scope_ and _dynamic scope_.
+
+In this blog post, I'd like to explain the differences between lexical scope
+and dynamic scope, and also explore some of the history behind those ideas. In
+a subsequent post, I'll discuss scoping in R and why it can be confusing.
+
+<!-- more -->
+
+## What is scope?
+
+_Scope_ refers to the places in a program where a variable is visible and can
+be referenced.
+
+An interesting situation is when a function has free variables. Consider the
+example below:
+
+```r
```

Comment:
is this R syntax? Is it meant to be ambiguous whether `<-` could mean assignment?

Reply:
Yeah, `<-` is assignment in R. Originally I had more R examples.

I'll clarify the description below to actually say assignment.

(Or I could use `x = 1` because that's also valid R. ...along with `1 -> x`.)

Reply:
I'm not sure I was clear. In every language I know, the binding of `x` from `x <- 1` is in scope for the `x <- 2` line. As someone who has never written R, but thinks `<-` looks vaguely like mutation, `x <- 2` could either mean (1) imperatively update the value of `x` from `1` to `2`, or (2) introduce a new binding of `x`, potentially shadowing the earlier one.

If the answer is (1), the example is unambiguous and doesn't really say anything about lexical vs. dynamic scope. It should just yield `2`.

p.s. it seems like we understand "assignment" to mean different things. I think of it as imperative update of mutable state, whereas you seem to also use it to mean the association of a name (immutably) with a value, which I would call "binding".

Reply:
Thanks for the clarification. In this specific example, the answer is (2): `x <- 2` does not update the previous value of `x`; it introduces a new binding for `x` that shadows the previous one.

But the ambiguity here is that `<-` is called the assignment operator in R, and it can also do updates:

```r
x <- 1   # creates a new variable
x        # value is 1
f <- function() {
  x <- 2 # creates a new variable
  x      # value is 2
  x <- 3 # updates the variable created 2 lines above
  x      # value is 3
}
x        # value is 1, refers to the variable created on the first line
```

`<-` works by searching the current function's scope for a variable to update. If a variable is found, it updates it. Otherwise, it creates a new binding* in the current scope. `<-` can only mutate the current scope.

I think I'll rewrite this example to not use R, and I'll include a discussion on `<-` in the next blog post.

*Maybe I need a different word, because these "bindings" are mutable. Mapping?

Reply:
I ended up keeping the R syntax, but clarifying what was going on in the prose.

```diff
-In this snippet, we define a variable `x` at the top level and two functions
-`f` and `g`. The body of `f` uses the parameter `a`, but also the free variable
-`x`. On line 7, we call `g`, which in turn calls `f` and returns that result.
+On line 1, we create a mapping for `x` with value `1`. On line 2, we define a
+function `f` whose body uses the parameter `a`, but also the free variable `x`.
+On line 3, we define a function `g`, whose body creates a new mapping for `x`
+with value `2`, and then calls `f(0)`. (Note that line 4 does not update the
+mapping created on line 1.) Finally, on line 7, we call `g()`.
 
-What value does `g` return when it is called? On line 2, what value should the
-free variable `x` be bound to?
+What value does `g` return when it is called? What mapping does the free
+variable `x` on line 2 refer to? Does it refer to the mapping on line 1 that
+was visible when `f` was defined? Or does it refer to the mapping on line 4
+that was created just before `f` was called?
```